server/src/graql/query/pattern/statement/StatementInstance.java

```diff
@@ -0,0 +1,371 @@
+        if (getProperty(IsaProperty.class).isPresent()) {
+            return getProperty(IsaProperty.class).get().toString();
+
+        } else if (getProperty(IsaExplicitProperty.class).isPresent()) {
+            return getProperty(IsaExplicitProperty.class).get().toString();
+
+        } else {
+            return "";
+        }
+    }
+
+    String hasSyntax() {
+        return this.properties().stream()
+                .filter(p -> p instanceof HasAttributeProperty)
+                .map(VarProperty::toString)
+                .collect(joining(Query.Char.COMMA_SPACE.toString()));
+    }
+
+    /**
+     * A Graql statement describe a Thing, which is the super type of an Entity,
+     * Relation and Attribute
+     */
+    public static class StatementThing extends StatementInstance {
+
+        public StatementThing(Variable var) {
+            this(var, new LinkedHashSet<>(), Sign.POSITIVE);
+        }
+
+        private StatementThing(Statement statement) {
+            this(statement.var(), statement.properties(), statement.sign());
+        }
+
+        StatementThing(Variable var, LinkedHashSet<VarProperty> properties, Sign sign) {
+            super(var, properties, sign);
+        }
+
+        public static StatementThing create(Statement statement, VarProperty varProperty) {
+            if (statement instanceof StatementThing) {
+                return ((StatementThing) statement).addProperty(varProperty);
+
+            } else if (!(statement instanceof StatementInstance)
+                    && !(statement instanceof StatementType)) {
+                return new StatementThing(statement).addProperty(varProperty);
+
+            } else {
+                throw illegalArgumentException(statement, varProperty);
+            }
+        }
+
+        @Override
+        public StatementThing negate() {
+            Sign negated = isPositive() ? Sign.NEGATIVE : Sign.POSITIVE;
+            return new StatementThing(var(), properties(), negated);
+        }
+
+        @CheckReturnValue
+        StatementThing addProperty(VarProperty property) {
+            validateNonUniqueOrThrow(property);
+            LinkedHashSet<VarProperty> newProperties = new LinkedHashSet<>(this.properties());
+            newProperties.add(property);
+            return new StatementThing(this.var(), newProperties, this.sign());
+        }
+
+        private String thingSyntax() {
+            if (!isaSyntax().isEmpty()) {
+                return isaSyntax();
+            } else if (getProperty(IdProperty.class).isPresent()) {
+                return getProperty(IdProperty.class).get().toString();
+
+            } else if (getProperty(NeqProperty.class).isPresent()) {
+                return getProperty(NeqProperty.class).get().toString();
+
+            } else {
+                return "";
+            }
+        }
+
+        @Override @SuppressWarnings("Duplicates")
+        public String toString() {
```

Comment:
we have three copies of this tostring

Reply:
they're not the same, I may or may not modularise them when I change Statement classes to be strict typing.